src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/X500DistinguishedNameBuilder.cs

```diff
@@ -0,0 +1,479 @@
+        /// </remarks>
+        public void AddStateOrProvinceName(string stateOrProvinceName)
+        {
+            // ITU T-REC X.520 Annex A:
+            // id-at-stateOrProvinceName
+            // WITH SYNTAX UnboundedDirectoryString
+
+            ArgumentException.ThrowIfNullOrEmpty(stateOrProvinceName);
+            EncodeComponent(Oids.StateOrProvinceName, stateOrProvinceName, UniversalTagNumber.UTF8String);
+        }
+
+        /// <summary>
+        /// Adds a domain component attribute.
+        /// </summary>
+        /// <param name="domainComponent">The domain component to add.</param>
+        /// <exception cref="ArgumentNullException">
+        /// <paramref name="domainComponent" /> is <see langword="null" />.
+        /// </exception>
+        /// <exception cref="ArgumentException">
+        /// <paramref name="domainComponent" /> is empty.
+        /// </exception>
+        /// <remarks>
+        /// This encodes an attribute with the OID 0.9.2342.19200300.100.1.25 as an IA5String.
+        /// </remarks>
+        public void AddDomainComponent(string domainComponent)
+        {
+            // RFC 5912
+            // id-domainComponent      AttributeType ::=
+            //  { itu-t(0) data(9) pss(2342) ucl(19200300) pilot(100)
+            //  pilotAttributeType(1) 25 }
+            //  at-domainComponent ATTRIBUTE ::= {TYPE IA5String
+            //  IDENTIFIED BY id-domainComponent }
+
+            ArgumentException.ThrowIfNullOrEmpty(domainComponent);
+            EncodeComponent(Oids.DomainComponent, domainComponent, UniversalTagNumber.IA5String);
+        }
+
+        /// <summary>
+        /// Builds an <see cref="X500DistinguishedName" /> that represents the encoded attributes.
+        /// </summary>
+        /// <returns>
+        /// An <see cref="X500DistinguishedName" /> that represents the encoded attributes.
+        /// </returns>
+        public X500DistinguishedName Build()
+        {
+            AsnWriter writer = new AsnWriter(AsnEncodingRules.DER);
+
+            using (writer.PushSequence())
+            {
+                foreach (byte[] component in _encodedComponents)
+                {
+                    writer.WriteEncodedValue(component);
+                }
+            }
+
+            byte[] rented = CryptoPool.Rent(writer.GetEncodedLength());
+            int encoded = writer.Encode(rented);
+            X500DistinguishedName name = new X500DistinguishedName(rented.AsSpan(0, encoded));
+            CryptoPool.Return(rented, clearSize: 0); // Distinguished Names do not contain sensitive information.
+            return name;
+        }
+
+        private void EncodeComponent(
+            string oid,
+            string value,
+            UniversalTagNumber stringEncodingType,
+            [CallerArgumentExpression("value")] string? paramName = null)
+        {
+            AsnWriter writer = new AsnWriter(AsnEncodingRules.DER);
```

Comment:
I thought about this... and got a little unsure from exception handling.

Consider if someone did:

```C#
X500DistinguishedNameBuilder builder = new();

try
{
    builder.AddEmailAddress(<something that is not IA5>);
}
catch {}

// What is the state of the shared writer here?
```

If the writer is shared, we might end up with things left-over in the writer like the pushed SET OF+SEQUENCE OF.

_Kevin, is that something that really matters?_

I... think... so. The shared writer should not be in an invalid state after exceptions.

_Okay Kevin, but we can call `Clear` *before* every use to make sure we have a clean slate instead of trying to clean up after ourselves_

We could... and we need to make a test to make sure the builder is still in a valid state after an exception... and the first `Reset` would effectively a no-op if we do them up-front... but wait doesn't `Reset` internally zero buffers since AsnWriter doesn't know if it's dealing with sensitive data?

So I just ended up creating a new writer every time. If we think it's wasteful, I wouldn't mind doing the work.

Reply:
AsnWriter.Clear() does zero out the contents of the buffer... but only as much as it has written.  Since it doesn't use an array pool, it's always fetching a zeroed 1k array for a new one; so holding one instance and calling Clear is probably less overall work.

But, yeah, Clear-first (like ERR_clear_error() first) is an interesting, and useful, observation/pattern.